Objective-C/CBLDatabase.mm

```diff
@@ -790,13 +814,107 @@ - (void) freeC4DB {
 }
 
 
-#pragma mark - DOCUMENT SAVE AND CONFLICT RESOLUTION
+#pragma mark - DOCUMENT SAVE AND CONFLICT HANDLING
+
+
+- (BOOL) saveDocument: (CBLDocument*)document
+   concurrencyControl: (CBLConcurrencyControl)concurrencyControl
+           asDeletion: (BOOL)deletion
+                error: (NSError**)outError
+{
+    if (deletion && !document.revID) {
+        [NSException raise: NSInvalidArgumentException
+                    format: @"Do not allow to delete a newly created document "
+         "that has not been saved into the database."];
+        return NO;
+    }
+    
+    CBL_LOCK(self) {
+        if (![self prepareDocument: document error: outError])
+            return NO;
+        
+        // Begin a db transaction:
+        C4Transaction transaction(_c4db);
+        if (!transaction.begin())
+            return convertError(transaction.error(), outError);
+        
+        C4Document* curDoc = nil;
+        if (deletion) {
+            // Check existing, NO-OPS if the document doesn't exist:
+            CBLStringBytes bDocID(document.id);
+            C4Error err;
+            curDoc = c4doc_get(_c4db, bDocID, true, &err);
+            if (!curDoc) {
+                if (err.code == kC4ErrorNotFound)
+                    return YES;
+                return convertError(err, outError);
+            }
+        }
+        
+        C4Document* newDoc = nil;
+        if (![self saveDocument: document into: &newDoc
+               withBaseDocument: nil asDeletion: deletion error: outError]) {
+            if (curDoc)
+                c4doc_free(curDoc);
+            return NO;
+        }
+        
+        if (newDoc) {
+            if (curDoc)
+                c4doc_free(curDoc);
+            
+            // Save succeeded; now commit the transaction:
```

Comment:
These lines (867-873) are pretty much identical to 904-909. Could you refactor to avoid the duplication? Like, take these lines out and instead wrap everything from 876-902 in `if (!newDoc) {...}`.